### PR TITLE
Support complex | feat(torchlib)

### PR DESCRIPTION
**Pull request by titaiwangms**
Adapt onnx-script to complex dtype.
The main work seems to be in converter side.

~~This is the first draft for discussion.~~

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/graph_building.py [hunk 1]
```diff
@@ -137,6 +138,14 @@ def dtype(self):
     def dtype(self, dtype: torch.dtype):
         self._torch_value.setType(self._torch_value.type().with_dtype(dtype))
 
+    @property
+    def is_complex(self):
+        return self._is_complex
+
+    @is_complex.setter
+    def is_complex(self, is_complex: bool):
+        self._is_complex = is_complex
```
Comment (titaiwangms) [inline onnxscript/function_libs/torch_lib/graph_building.py:147]: Converter usage

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py [hunk 2]
```diff
@@ -734,6 +735,11 @@ def _where_input_wrangler(
 
 
 SKIP_XFAIL_SUBTESTS: tuple[ops_test_common.DecorateMeta, ...] = (
+    skip(
+        "abs_complex",
+        matcher=lambda sample: sample.input.dtype == torch.float32,
```
Comment (titaiwangms) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:740]: I wonder how general should we make on dtype in ops_test.py right now given that we only test float32.

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py [hunk 3]
```diff
@@ -254,6 +254,9 @@ def duplicate_opinfo_for_prims(
 
 def convert_tensor_to_numpy(input: Any) -> Any:
     if isinstance(input, torch.Tensor):
+        if input.dtype in (torch.complex64, torch.complex32):
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test_common.py:257]: Looks like there's `torch.is_complex` for us

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_data.py:740]: Potentially leverage the dtypes field (https://github.com/titaiwangms/onnx-script/blob/1f6d993ba8a19f5213c015e99a3081a12dfb1777/onnxscript/tests/function_libs/torch_lib/ops_test_common.py#L86), and create a constant of something like `REAL_TYPES`. In skip we can use `skip(..., dtypes=REAL_TYPES)`

`REAL_TYPES = (*BOOL_TYPES, *INT_TYPES, *FLOAT_TYPES)`

Alternatively, we can create a new mapping (TESTED_COMPLEX_OPS) and parameterize the test only for complex types. We can duplicate this test

https://github.com/titaiwangms/onnx-script/blob/1f6d993ba8a19f5213c015e99a3081a12dfb1777/onnxscript/tests/function_libs/torch_lib/ops_test.py#L322-L330

like so

```python
    @ops_test_common.add_decorate_info(
        ops_test_data.OPS_DB,
        "TestOutputConsistencyFullGraph",
        "test_output_match_opinfo_complex_types_",  # Update here
        skip_or_xfails=ops_test_data.EXPECTED_SKIPS_OR_FAILS,
    )
    @common_device_type.ops(  # type: ignore[misc]
        [info for info in ops_test_data.OPS_DB if info.name in ops_test_data.TESTED_COMPLEX_OPS],  # Update here
        allowed_dtypes=COMPLEX_TYPES,  # Update here
    )
    def test_output_match_opinfo_complex_types_(  # Update here
        self, device: str, dtype: torch.dtype, op: opinfo_core.OpInfo
    ):
        """Base test method for testing each op by running the full ONNX graph."""
        run_test_output_match(self, device, dtype, op, ops_test_common.graph_executor)
```

This way we may not need as many xfails, and we don't need to duplicate opinfos.

Same thing can be done for the test in `TestOutputConsistencyEager`

File: onnxscript/tests/function_libs/torch_lib/ops_test.py [hunk 4]
```diff
@@ -192,6 +196,12 @@ def run_test_output_match(
                     input_onnx, kwargs_onnx = input_wrangler(input_onnx, kwargs_onnx)
                 torch_output = op(*inputs, **cpu_sample.kwargs)
 
+                if (
+                    isinstance(torch_output, torch.Tensor)
+                    and torch_output.dtype == torch.complex64
```
Comment (titaiwangms) [inline onnxscript/tests/function_libs/torch_lib/ops_test.py:201]: Use torch.is_complex

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py (hunk 2 above)
Comment (titaiwangms) [reply onnxscript/tests/function_libs/torch_lib/ops_test_data.py:740]: Very nice!

File: onnxscript/function_libs/torch_lib/graph_building.py [hunk 5]
```diff
@@ -137,6 +138,14 @@ def dtype(self):
     def dtype(self, dtype: torch.dtype):
         self._torch_value.setType(self._torch_value.type().with_dtype(dtype))
 
+    @property
+    def is_complex(self):
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/graph_building.py:142]:
```suggestion
    def is_complex(self) -> bool:
```

nit

File: onnxscript/tests/function_libs/torch_lib/ops_test.py [hunk 6]
```diff
@@ -149,20 +161,24 @@ def run_test_output_match(
...
+    onnx_function_and_wrangler = tested_op_mapping[op.name]
     # Obtain the input_wrangler that manipulates the OpInfo inputs
     # to match the aten operator signature
     # An example is nn.functional.upsample_nearest2d, which has a different signature
     # than the aten operator upsample_nearest2d
     onnx_function, input_wrangler = _split_function_and_wrangler(onnx_function_and_wrangler)
-    if not ops_test_common.dtype_op_schema_compatible(dtype, onnx_function.op_schema):
+    if (
+        not ops_test_common.dtype_op_schema_compatible(dtype, onnx_function.op_schema)
+        and dtype != torch.complex64
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test.py:180]: is_complex

Comment (titaiwangms) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py]: torch.is_complex only works on tensors. I changed this to COMPLEX_TYPES we define in the file

File: onnxscript/tests/function_libs/torch_lib/ops_test.py [hunk 7]
```diff
@@ -177,7 +177,7 @@ def run_test_output_match(
     onnx_function, input_wrangler = _split_function_and_wrangler(onnx_function_and_wrangler)
     if (
         not ops_test_common.dtype_op_schema_compatible(dtype, onnx_function.op_schema)
-        and dtype != torch.complex64
+        and dtype not in COMPLEX_TYPES
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test.py:180]: Actually, We shouldn’t need this right ?

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:180]: Nvm

Comment (titaiwangms) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:180]: Oh yeah, The schema checks has to be bypassed

**Pull request merged**